Issue95 - Added USB + HW serial support for 4630 and 461x boards

svelmurugan92 opened the pull request:
#95 #70 Added USB + HW serial support in 4630 and 461x board. 
We can able to choose this option from Arduino IDE serial interface menu "USB+HW Serial (USART1 and 2 if available)"

terrillmoore commented on boards.template:
```diff
@@ -205,6 +205,8 @@ $section xserial
 {{board}}.menu.xserial.none.build.xSerial=-DNO_HWSERIAL
 #{{board}}.menu.xserial.all=All (up to 10)
 #{{board}}.menu.xserial.all.build.xSerial=-DALL_HWSERIAL
+{{board}}.menu.xserial.usbhwserial=USB+HW Serial (USART1 and 2 if available)
```
Can we add spaces around the "+"?
```suggestion
{{board}}.menu.xserial.usbhwserial=USB + HW Serial (USART1 and 2 if available)
```

terrillmoore replied:
Under what circumstances are USART1 and/or 2 not available?  If always available, remove "if available".

terrillmoore commented on boards.template:
```diff
@@ -217,6 +219,8 @@ $section xserial
 {{board}}.menu.xserial.none.build.xSerial=-DNO_HWSERIAL
 #{{board}}.menu.xserial.all=All (up to 10)
 #{{board}}.menu.xserial.all.build.xSerial=-DALL_HWSERIAL
+{{board}}.menu.xserial.usbhwserial=USB+HW Serial (USART1 and 2 if available)
```
```suggestion
{{board}}.menu.xserial.usbhwserial=USB + HW Serial (USART1 and 2 if available)
```

terrillmoore commented on boards.template:
```diff
@@ -205,6 +205,8 @@ $section xserial
 #{{board}}.menu.xserial.all.build.xSerial=-DALL_HWSERIAL
+{{board}}.menu.xserial.usbhwserial=USB+HW Serial (USART1 and 2 if available)
+{{board}}.menu.xserial.usbhwserial.build.xSerial={build.usb_flags} -DUSB_TWO_HWSERIAL
 #{{board}}.menu.xserial.third=Enable first third (USART1 to 3 if available)
```
Under what circumstances are USART 2 or USART 3 not availalbe? I think this is chip specific, and we only have one kind of chip, so "if available" doesn't apply?  If I'm missing something, please explain.

svelmurugan92 replied:
In our board USART1 and USART2 are only available and enabled in PeripheralPins.c. [variants/CATENA_4630/PeripheralPins.c](url)
https://github.com/mcci-catena/Arduino_Core_STM32/blob/4e62436fa139793a8924029ed1d5c2ed1f1b9e64/variants/CATENA_4630/PeripheralPins.c#L141
```//*** SERIAL ***

#ifdef HAL_UART_MODULE_ENABLED
const PinMap PinMap_UART_TX[] = {
//    {PA_0,  USART4,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_USART4)},
//    {PA_2,  LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
    {PA_2,  USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_NOPULL, GPIO_AF4_USART2)},
    {PA_9,  USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_NOPULL, GPIO_AF4_USART1)},
//    {PA_14, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
//    {PA_14, USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_USART2)},
//    {PB_3,  USART5,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_USART5)},
//    {PB_6,  USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF0_USART1)},
//    {PB_10, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_LPUART1)},
//    {PB_11, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_LPUART1)},
//    {PC_1,  LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
    {NC,    NP,    0}
};
#endif

#ifdef HAL_UART_MODULE_ENABLED
const PinMap PinMap_UART_RX[] = {
//    {PA_1,  USART4,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_USART4)},
//    {PA_3,  LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
    {PA_3,  USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_NOPULL, GPIO_AF4_USART2)},
    {PA_10, USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_NOPULL, GPIO_AF4_USART1)},
//    {PA_13, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
//    {PA_15, USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_USART2)},
//    {PB_4,  USART5,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_USART5)},
//    {PB_7,  USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF0_USART1)},
//    {PB_10, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_LPUART1)},
//    {PB_11, LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_LPUART1)},
//    {PC_0,  LPUART1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_LPUART1)},
    {NC,    NP,    0}
};
#endif```

USART 3 is not available.

terrillmoore replied:
Given this, we should remove (or rename) this option. I'm updating to accommodate.

svelmurugan92 replied:
Small correction: Only in 4630 and 4801 (To support PMS7003 and RS485 transceiver respectively) we enabled extra HW serial USART2 . In the 461x and 4551 we enabled only one HW serial USART1

terrillmoore commented on variants/CATENA_461x/variant.h:
```diff
@@ -122,12 +122,6 @@ enum {
 
 // UART Definitions
-// Define here Serial instance number to map on Serial generic name
-#define SERIAL_UART_INSTANCE    1 //ex: 2 for Serial2 (USART2)
```
This is a change. Formerly, `SERIAL_UART_INSTANCE` was always defined, even if USB.  Now, if USB, it's not defined. This is a breaking change for any code that depends on it being defined.  I think we should define it in all cases. Ditto for `PIN_SERIAL_RX` and `PIN_SERIAL_TX` -- formerly, these were always defined, even if not sensible. I would define them in each brnch of the `#ifdef` below.

terrillmoore replied:
I took care of this in the new header file.

terrillmoore commented on variants/CATENA_461x/variant.h:
```diff
@@ -152,14 +146,24 @@ enum {
 # define SERIAL_PORT_MONITOR  SerialNo
 #else
+// Define here Serial instance number to map on Serial generic name
+#define SERIAL_UART_INSTANCE    1
```
```suggestion
# define SERIAL_UART_INSTANCE    1
```

terrillmoore commented on variants/CATENA_461x/variant.h:
```diff
@@ -152,14 +146,24 @@ enum {
 // SERIAL_PORT_HARDWARE_OPEN  Hardware serial ports which are open for use.  Their RX & TX
 //                            pins are NOT connected to anything by default.
-#ifdef USBCON
+#if defined (USBCON) && defined (USB_TWO_HWSERIAL)
```
```suggestion
#if defined(USBCON) && defined(USB_TWO_HWSERIAL)
```
We don't uses spaces here in MCCI code.

terrillmoore commented on variants/CATENA_461x/variant.h:
```diff
@@ -152,14 +146,24 @@ enum {
+# define SERIAL_PORT_MONITOR  SerialUSB
+# define HAVE_HWSERIAL1
+# define HAVE_HWSERIAL2
+#elif defined (USBCON)
```
```suggestion
#elif defined(USBCON) /* && ! defined(USB_TWO_HWSERIAL)
```

But this organization is getting quite convoluted because it's the wrong way to do this. See upcoming main comment.

terrillmoore commented on variants/CATENA_461x/variant.h:
```diff
@@ -152,14 +146,24 @@ enum {
 # define SERIAL_PORT_MONITOR  SerialUSB
 #elif defined(NO_HWSERIAL)
 # define SERIAL_PORT_MONITOR  SerialNo
 #else
```
Need a comment on the `#else`.
```suggestion
#else /* ! defined(USBCON) && !defined(NO_HWSERIAL) */
```

terrillmoore commented on boards.txt:
```diff
@@ -93,6 +93,8 @@ mcci_catena_4551.menu.xserial.usb.build.xSerial={build.usb_flags} -DNO_HWSERIAL
 mcci_catena_4551.menu.xserial.none.build.xSerial=-DNO_HWSERIAL
 #mcci_catena_4551.menu.xserial.all=All (up to 10)
 #mcci_catena_4551.menu.xserial.all.build.xSerial=-DALL_HWSERIAL
+mcci_catena_4551.menu.xserial.usbhwserial=USB+HW Serial (USART1 and 2 if available)
```
We obviously edited boards.txt rather than regenerating.

It is not permitted ever to edit boards.txt. It must be generated. See the instructions.  This means you must update the templates for the other boards.

I will do that.

terrillmoore commented:
I take it back, I see we did regenerate boards.txt; I fixed several things, centralized things, and regenerated. Key change:

1) default config is now what people will mostly want.
2) a header file decodes ARDUINO_BSP_CONFIG_SERIAL_MASK to set up the configs for all variants.

Please test this, thanks.

svelmurugan92 commented on variants/mcci_common_uart.h:
```diff
@@ -175,13 +176,15 @@ is correct.
 #endif
 
 // post-condition: PIN_SERIAL_RX and _TX are defined.
+//if ! ARDUINO_BSP_SERIAL_USB_IS_ENABLED() &&
```
@terrillmoore while USB is enable if we redefine PIN_SERIAL_RX and PIN_SERIAL_TX cause below build issue. So, before defining these, i checked for USB enable. Please provide your views
```` In file included from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\wiring.h:49:0,

                 from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\Arduino.h:32,

                 from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp:30:

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp: In constructor 'HardwareSerial::HardwareSerial(void*)':

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino/stm32/usb_serial.h:73:17: error: comparison between distinct pointer types 'HardwareSerial*' and 'USBSerial*' lacks a cast

 # define Serial SerialUSB

                 ^

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp:160:16: note: in expansion of macro 'Serial'

   if (this == &Serial)

                ^~~~~~

exit status 1
Error compiling for board MCCI Catena 4630.```

svelmurugan92 commented on variants/mcci_common_uart.h:
```diff
@@ -176,15 +176,15 @@ is correct.
 
 // post-condition: PIN_SERIAL_RX and _TX are defined.
-//if ! ARDUINO_BSP_SERIAL_USB_IS_ENABLED() &&
+if ! ARDUINO_BSP_SERIAL_USB_IS_ENABLED() &&
```
@terrillmoore while USB is enable if we redefine PIN_SERIAL_RX and PIN_SERIAL_TX cause below build issue. So, before defining these, i checked for USB enable. Please provide your views
```` In file included from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\wiring.h:49:0,

                 from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\Arduino.h:32,

                 from C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp:30:

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp: In constructor 'HardwareSerial::HardwareSerial(void*)':

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino/stm32/usb_serial.h:73:17: error: comparison between distinct pointer types 'HardwareSerial*' and 'USBSerial*' lacks a cast

 # define Serial SerialUSB

                 ^

C:\Users\velmurugan\Documents\Arduino\hardware\MCCI\stm32\cores\arduino\HardwareSerial.cpp:160:16: note: in expansion of macro 'Serial'

   if (this == &Serial)

                ^~~~~~

exit status 1
Error compiling for board MCCI Catena 4630.```

terrillmoore replied:
This is a conceptual error in `HardwareSerial.cpp`, not in the header file. It should be comparing `(void *) this` to `(void *)&Serial` (in the context of this check) [it's only trying to compare pointers].  They forgot that the type of Serial could be any valid derived type, and so they can't compare pointers without removing the type (or changing both pointers to the base type). Given what they're trying to learn here, casting to `void*` in both places is appropriate and portable.

So please revert the change on the definitions, fix this problem, and find out what else is broken.... You can file a separate issue for this ("HardwareSerial mistakenly assumes Serial is of type HardwareSerial"), but fix it on the Issue95 branch. Just make sure the commit message has "Fix #NNN: ..." (whatever the issue number is) at the beginning.

terrillmoore replied:
I'm going to file the issue and fix this now.

terrillmoore replied:
See #99

svelmurugan92 commented:
> I take it back, I see we did regenerate boards.txt; I fixed several things, centralized things, and regenerated. Key change:
> 
> 1. default config is now what people will mostly want.
> 2. a header file decodes ARDUINO_BSP_CONFIG_SERIAL_MASK to set up the configs for all variants.
> 
> Please test this, thanks.

Thank you for the update. 
We build the code and fixed few build issues. Committed the code changes.

terrillmoore commented:
I've pushed new changes to address #99. Please update and test.

svelmurugan92 commented:
@terrillmoore  We tested the changes. 
Able to access the USB and HW serial. Works fine. 
We can merge the changes

dhineshkumarmcci commented:
We also removed the `build_board_old_flags` for the new board Catena 4630 #94 . This is fix is done with reference to issue #49

terrillmoore commented:
Let's get @chwon64's review, then merge and start the release process.

chwon64 commented on boards.template:
```diff
@@ -201,32 +201,29 @@ $section xserial
 {{board}}.menu.xserial.generic=Generic Serial
 {{board}}.menu.xserial.none=No Serial
+{{board}}.menu.xserial.usbhwserial.build.xSerial={build.usb_flags} -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x7
 {{board}}.menu.xserial.usb.build.xSerial={build.usb_flags} -DNO_HWSERIAL
```
I think you need to use ARDUINO_BSP_CONFIG_SERIAL_MASK instead of NO_HWSERIAL.

svelmurugan92 replied:
@chwon64 Thank you for the review. We will change this

chwon64 commented on boards.txt:
```diff
@@ -812,15 +806,14 @@ mcci_catena_4630.menu.upload_method.STLink.upload.tool=stlink_upload
 mcci_catena_4630.menu.xserial.generic=Generic Serial
 mcci_catena_4630.menu.xserial.none=No Serial
+mcci_catena_4630.menu.xserial.usbhwserial.build.xSerial={build.usb_flags} -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x7
 mcci_catena_4630.menu.xserial.usb.build.xSerial={build.usb_flags} -DNO_HWSERIAL
```
Same here. Please check all place where use NO_HWSERIAL.

svelmurugan92 commented:
@chwon64 Pushed the changes. Please review once.

chwon64 commented on boards.template:
```diff
@@ -214,7 +214,7 @@ $section xserial
 {{board}}.menu.xserial.none=No Serial
 {{board}}.menu.xserial.usbhwserial.build.xSerial={build.usb_flags} -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x7
-{{board}}.menu.xserial.usb.build.xSerial={build.usb_flags} -DNO_HWSERIAL
+{{board}}.menu.xserial.usb.build.xSerial={build.usb_flags} -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x0
```
It should be 1 for USB.

svelmurugan92 replied:
I think, It is USB only option. So provide {build.usb_flags} and NO_HWSERIAL flag mask -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x0.
Maybe in platform.txt we have to add the flag -DARDUINO_BSP_CONFIG_SERIAL_MASK=0x1 in {build.usb_flags}. 

[https://github.com/mcci-catena/Arduino_Core_STM32/blob/4395e6b81ad61a52ce55a7362fe5ce3566f9f522/platform.txt#L82](url)

Now ARDUINO_BSP_CONFIG_SERIAL_MASK 1 is defined using USBCON flag here.
[https://github.com/mcci-catena/Arduino_Core_STM32/blob/4395e6b81ad61a52ce55a7362fe5ce3566f9f522/variants/mcci_common_uart.h#L97](url)

Please provide your views

chwon64 replied:
No, just change value to 1 because all other place use 1.

svelmurugan92 commented:
Not sure. Its normal. 
While i generate the boards.txt using below command
> bright generate_boards.bri boards.template boards.json > boards.txt 

It generate like a bin file. so, github compare not works.

terrillmoore commented:
> 
> 
> Not sure. Its normal.
> While i generate the boards.txt using below command
> 
> > bright generate_boards.bri boards.template boards.json > boards.txt
> 
> It generate like a bin file. so, github compare not works.

Somehow the generation got messed up. I will see if I can fix.

terrillmoore commented:
> > Not sure. Its normal.
> > While i generate the boards.txt using below command
> > > bright generate_boards.bri boards.template boards.json > boards.txt
> > 
> > It generate like a bin file. so, github compare not works.
> 
> Somehow the generation got messed up. I will see if I can fix.

Yup. The file was garbage. I regenerated and it's fine. Must have been a typo when you ran the command.
